pcdsdevices/gon.py

```diff
@@ -236,23 +252,351 @@ class Kappa(BaseInterface, Device):
+        return e_phi
+
+    @real_position_argument
+    def k_to_e(self, real_pos=None):
+        """
+        Convert from native kappa coordinates to spherical coordinates.
+
+        If a parameter is left as None, use the live value.
+
+        Parameters
+        ----------
+        eta : number
+            Eta motor position.
+        kappa : number
+            Kappa motor position.
+        phi : number
+            Phi motor position.
+
+        Returns
+        -------
+        coordinates : tuple
+            Spherical coordinates.
+        """
+        real_pos = self.RealPosition(*real_pos)
+
+        if not real_pos.eta and real_pos.eta != 0:
+            real_pos.eta = self.eta_position
+        if not real_pos.kappa and real_pos.kappa != 0:
+            real_pos.kap = self.kappa_position
+        if not real_pos.phi and real_pos.phi != 0:
+            real_pos.phi = -self.phi_position
+
+        kappa_ang = self.kappa_ang * np.pi / 180
+
+        delta = np.arctan(np.tan(real_pos.kappa * np.pi /
+                                 180 / 2.0) * np.cos(kappa_ang))
+        e_eta = -real_pos.eta * np.pi / 180 - delta
+        e_chi = 2.0 * np.arcsin(np.sin(real_pos.kappa * np.pi /
+                                       180 / 2.0) * np.sin(kappa_ang))
+        e_phi = real_pos.phi * np.pi / 180 - delta
+
+        e_eta = e_eta * 180 / np.pi
+        e_chi = e_chi * 180 / np.pi
+        e_phi = e_phi * 180 / np.pi
+        # return e_eta, e_chi, e_phi
+        return self.PseudoPosition(e_eta=e_eta, e_chi=e_chi, e_phi=e_phi)
+
+    @pseudo_position_argument
+    def e_to_k(self, pseudo_pos):
+        """
+        Convert from spherical coordinates to the native kappa coordinates.
+
+        If a parameter is left as None, use the live value.
+
+        Parameters
+        ----------
+        e_eta : number
+        e_chi : number
+        e_phi : number
+
+        Returns
+        -------
+        coordinates : tuple
+            Native kappa coordinates.
+        """
+        pseudo_pos = self.PseudoPosition(*pseudo_pos)
+        if not pseudo_pos.e_eta and pseudo_pos.e_eta != 0:
+            pseudo_pos.e_eta = self.e_eta_coord
+        if not pseudo_pos.e_chi and pseudo_pos.e_chi != 0:
+            pseudo_pos.e_chi = self.e_chi_coord
+        if not pseudo_pos.e_phi and pseudo_pos.e_phi != 0:
+            pseudo_pos.e_phi = self.e_phi_coord
+
+        kappa_ang = self.kappa_ang * np.pi / 180
+
+        delta = np.arcsin(-np.tan(pseudo_pos.e_chi * np.pi / 180 /
+                                  2.0) / np.tan(kappa_ang))
+        k_eta = -(pseudo_pos.e_eta * np.pi / 180 - delta)
+        k_kap = 2.0 * np.arcsin(np.sin(pseudo_pos.e_chi * np.pi /
+                                       180 / 2.0) / np.sin(kappa_ang))
+        k_phi = pseudo_pos.e_phi * np.pi / 180 - delta
+
+        k_eta = k_eta * 180 / np.pi
+        k_kap = k_kap * 180 / np.pi
+        k_phi = -k_phi * 180 / np.pi
+        # return k_eta, k_kap, k_phi
```

Comment:
I think i should leave these calculations alone without the `pseudo_pos` and `real_pos` and only handle that in the `forward` and `inverse` methods, by calling these functions?

Reply:
I think isolating the calculations is a good idea, especially since there are three "real motor" elements that are omitted.